hot-reload-analysis/src/main/kotlin/org/jetbrains/compose/reload/analysis/RuntimeInstructionTree.kt

```diff
@@ -254,3 +251,168 @@ private fun parseRuntimeInstructionTree(
         lastIndex = index
     ).toLeft()
 }
+
+
+private class MutableTree(
+    var group: ComposeGroupKey?,
+    var type: RuntimeScopeType,
+    val tokens: MutableList<RuntimeInstructionToken>,
+    val children: MutableList<MutableTree>,
+    var startIndex: Int,
+    var endIndex: Int,
+    val markers: MutableSet<Int> = mutableSetOf()
+) {
+    fun toRuntimeInstructionTree(): RuntimeInstructionTree = RuntimeInstructionTree(
+        group = this.group,
+        type = this.type,
+        tokens = this.tokens,
+        children = this.children.map { it.toRuntimeInstructionTree() },
+        startIndex = this.startIndex,
+        lastIndex = this.endIndex,
+    )
+}
+
+private fun MutableList<MutableTree?>.matchToken2Tree(
+    tokenIndex: Int,
+    tree: MutableTree?
+) {
+    if (this[tokenIndex] == null) {
+        this[tokenIndex] = tree
+    }
+}
+
+private fun linearParseRuntimeInstructionTree(
+    methodNode: MethodNode, tokens: List<RuntimeInstructionToken>
+): Either<RuntimeInstructionTree, Failure> {
+    if (tokens.isEmpty()) return Failure("empty tokens").toRight()
+
+    val tree2Parent = MutableList<MutableTree?>(tokens.size + 1) { null }
```

Comment:
When we do have a `MutableTree` class: Would you think we can also set the parent on this `MutableTree` directly instead of creating this index?